src/Microsoft.Health.Dicom.Core/Features/Workitem/WorkitemOrchestrator.cs

```diff
@@ -109,83 +107,69 @@ public async Task CancelWorkitemAsync(DicomDataset dataset, WorkitemMetadataStor
-                storeDicomDataset = await GetWorkitemBlobAsync(workitemInstanceIdentifier, cancellationToken).ConfigureAwait(false);
+                identifier = new WorkitemInstanceIdentifier(
+                    workitemMetadata.WorkitemUid,
+                    workitemMetadata.WorkitemKey,
+                    workitemMetadata.PartitionKey,
+                    watermarkEntry.Value.CurrentWatermark);
+                var storeDicomDataset = await GetWorkitemBlobAsync(identifier, cancellationToken).ConfigureAwait(false);
 
                 // update the procedure step state
                 var updatedDicomDataset = new DicomDataset(storeDicomDataset);
 
+                // Add/Update the procedure step state in the blob
+                var targetProcedureStepStateStringValue = targetProcedureStepState.GetStringValue();
+                updatedDicomDataset.AddOrUpdate(DicomTag.ProcedureStepState, targetProcedureStepStateStringValue);
+
                 // if there is a reason for cancellation, set it in the blob
                 dataset.TryGetString(DicomTag.ReasonForCancellation, out var cancellationReason);
                 if (!string.IsNullOrWhiteSpace(cancellationReason))
                 {
                     updatedDicomDataset.AddOrUpdate(DicomTag.ReasonForCancellation, cancellationReason);
                 }
 
-                // Add/Update the procedure step state in the blob
-                updatedDicomDataset.AddOrUpdate(DicomTag.ProcedureStepState, targetProcedureStepState.GetStringValue());
-
-                // update the workitem in the blob store
-                await StoreWorkitemBlobAsync(workitemInstanceIdentifier, updatedDicomDataset, cancellationToken).ConfigureAwait(false);
+                // store the blob with the new watermark
+                await StoreWorkitemBlobAsync(identifier, updatedDicomDataset, watermarkEntry.Value.NextWatermark, cancellationToken)
+                    .ConfigureAwait(false);
 
-                // Update the workitem tags in the store
-                var queryTags = await _workitemQueryTagService.GetQueryTagsAsync(cancellationToken).ConfigureAwait(false);
+                // Update the workitem procedure step state in the store
                 await _indexWorkitemStore
-                    .EndUpdateWorkitemAsync(workitemMetadata, updatedDicomDataset, queryTags, cancellationToken)
+                    .UpdateWorkitemProcedureStepStateAsync(
+                        workitemMetadata,
+                        watermarkEntry.Value.NextWatermark,
+                        targetProcedureStepStateStringValue,
+                        cancellationToken)
+                    .ConfigureAwait(false);
+
+                // Delete the blob with the old watermark
+                await TryDeleteWorkitemBlobAsync(identifier, watermarkEntry.Value.CurrentWatermark, cancellationToken)
```

Comment:
watermarkEntry.Value.CurrentWatermark this can be removed since the identified already has the current watermark